cloudinit/net/dhcp.py

```diff
@@ -267,101 +131,366 @@ def networkd_get_option_from_leases(keyname, leases_d=None):
     return None
 
 
-def parse_static_routes(rfc3442):
-    """parse rfc3442 format and return a list containing tuple of strings.
+class DhcpClient(abc.ABC):
+    client_name = ""
 
-    The tuple is composed of the network_address (including net length) and
-    gateway for a parsed static route.  It can parse two formats of rfc3442,
-    one from dhcpcd and one from dhclient (isc).
+    @classmethod
+    def kill_dhcp_client(cls):
+        subp.subp(["pkill", cls.client_name], rcs=[0, 1])
 
-    @param rfc3442: string in rfc3442 format (isc or dhcpd)
-    @returns: list of tuple(str, str) for all valid parsed routes until the
-              first parsing error.
+    @classmethod
+    def clear_leases(cls):
+        cls.kill_dhcp_client()
+        files = glob.glob("/var/lib/dhcp/*")
+        for file in files:
+            os.remove(file)
 
-    E.g.
-    sr=parse_static_routes("32,169,254,169,254,130,56,248,255,0,130,56,240,1")
-    sr=[
-        ("169.254.169.254/32", "130.56.248.255"), ("0.0.0.0/0", "130.56.240.1")
-    ]
+    @classmethod
+    def start_service(cls, dhcp_interface: str, distro):
+        distro.manage_service("start", cls.client_name, rcs=[0, 1])
 
-    sr2 = parse_static_routes("24.191.168.128 192.168.128.1,0 192.168.128.1")
-    sr2 = [
-        ("191.168.128.0/24", "192.168.128.1"), ("0.0.0.0/0", "192.168.128.1")
-    ]
+    @classmethod
+    def stop_service(cls, dhcp_interface: str, distro):
+        distro.manage_service("stop", cls.client_name, rcs=[0, 1])
 
-    Python version of isc-dhclient's hooks:
-       /etc/dhcp/dhclient-exit-hooks.d/rfc3442-classless-routes
-    """
-    # raw strings from dhcp lease may end in semi-colon
-    rfc3442 = rfc3442.rstrip(";")
-    tokens = [tok for tok in re.split(r"[, .]", rfc3442) if tok]
-    static_routes = []
-
-    def _trunc_error(cidr, required, remain):
-        msg = (
-            "RFC3442 string malformed.  Current route has CIDR of %s "
-            "and requires %s significant octets, but only %s remain. "
-            "Verify DHCP rfc3442-classless-static-routes value: %s"
-            % (cidr, required, remain, rfc3442)
+
+class IscDhclient(DhcpClient):
+    client_name = "dhclient"
+
+    def __init__(self):
+        self.dhclient_path = subp.which("dhclient")
+        if not self.dhclient_path:
+            LOG.debug(
+                "Skip dhclient configuration: No dhclient command found."
+            )
+            raise NoDHCPLeaseMissingDhclientError()
+
+    @staticmethod
+    def parse_dhcp_lease_file(lease_file):
+        """Parse the given dhcp lease file for the most recent lease.
```

Comment:
This returns a list of all leases. Let's use type annotations on the function definitions here to guide our expectations about params and return types
And let's also either:
* fix the logic to return only one lease and correct the one call-site in ephemeral.py to assume most recent lease
-- or --
* Fix the docstr header

Reply:
fixed in 5cd048f6311ae4c509d890d3dfb0320913b6252c

Reply:
Sorry, I must have accidentally dropped that in a rebase. Thanks!